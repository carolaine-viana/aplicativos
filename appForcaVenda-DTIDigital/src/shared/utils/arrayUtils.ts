/**
 * Use this to deal/manipulate arrays
 */
export function isArrayEmpty<T = null>(arr: Array<T> | null | undefined) {
  return arr && arr?.length === 0;
}

export function doesArrayHaveContent<T = null>(arr: Array<T> | null | undefined) {
  const isEmpty = isArrayEmpty(arr);
  if (isEmpty) {
    return false;
  }
  return arr?.some((i) => !!i);
}

export function filterArraObjectAnyValue<T>(
  arr: Array<T>,
  valueFilter: string,
  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  keysFilter?: Array<keyof T>
) {
  return arr.filter((item: T) => {
    const values = Object.values(item).map((value: string) => {
      if (typeof value !== 'object') {
        return String(value).toUpperCase();
      }
      return null;
    });
    return values.join(' ').includes(valueFilter.toUpperCase());
  });
}

export const addNoRepeatedString = (list: string[], newValue: string) => {
  const newList = list.filter((str) => str !== newValue || str.length === 0);
  newList.push(newValue);
  return newList;
};

export const mergArrayIntoStrings = (list: string[], separator = ';') => {
  if (!list || isArrayEmpty(list)) {
    return '';
  }

  const noEmptyStrings = list.filter((str) => str.length > 0);
  const mergedString =
    noEmptyStrings.length === 1 ? noEmptyStrings[0] : noEmptyStrings.join(separator);

  return mergedString;
};
