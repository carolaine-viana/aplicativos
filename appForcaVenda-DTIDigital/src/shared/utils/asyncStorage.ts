import AsyncStorage from '@react-native-community/async-storage';
import { Profiles, TRole } from '@shared/types';
import { addNoRepeatedString, getISOStringDate, mergArrayIntoStrings } from '.';

export const FV_TOKEN = '@fv/token';
export const FV_TOKEN_EXPIRATION = '@fv/tokenExpiration';
export const FV_USER_LOGIN = '@fv/rcaLogin';
export const FV_USER_ROLE = '@fv/userRole';
export const FV_FIRST_TUTO_MAP = '@fv/firstTutoMap';
export const FV_FEATURE_FLAGS = '@fv/featureFlags';
export const FV_COMPLETED_TUTORIALS = '@fv/completedTutorials';
export const FV_LAST_TIME_USER_SAW_TUTORIAL = '@fv/lastTimeUserSawTutorial';
export const FV_DATE_ACCEPTED_TERM_USE = '@fv/dateAcceptedTermUse';
export const FV_DATE_UPDATED_TERM_USE = '@fv/dateUpdatedTermUse';

export const clear = () => AsyncStorage.clear();

// TOKEN INFO
export const setToken = async (token: string) => {
  await AsyncStorage.setItem(FV_TOKEN, token);
};

export const getToken = async () => {
  const token = await AsyncStorage.getItem(FV_TOKEN);
  return token;
};

export const setTokenExpiration = async (expirationDate: Date) => {
  await AsyncStorage.setItem(FV_TOKEN_EXPIRATION, expirationDate.toString());
};

export const getTokenExpiration = async (): Promise<Date | number> => {
  const expirationTime = await AsyncStorage.getItem(FV_TOKEN_EXPIRATION);
  return expirationTime ? new Date(expirationTime) : 0;
};

// USER INFO
export const getUserLogin = async () => {
  const userLogin = await AsyncStorage.getItem(FV_USER_LOGIN);
  return userLogin;
};

export const setUserLogin = async (login: string) => {
  await AsyncStorage.setItem(FV_USER_LOGIN, login);
};

export const getUserRole = async (): Promise<TRole | null> => {
  const userRole = await AsyncStorage.getItem(FV_USER_ROLE);
  return userRole as TRole;
};

export const setUserRole = async (role: TRole) => {
  await AsyncStorage.setItem(FV_USER_ROLE, role);
};

export const isUserGV = async (): Promise<boolean> => {
  const userRole = await getUserRole();
  return userRole === Profiles.GV;
};

// FEATURE FLAG
export const setFeatureFlags = async (featureFlags: string) => {
  await AsyncStorage.setItem(FV_FEATURE_FLAGS, featureFlags);
};

export const getFeatureFlags = async () => {
  const flags = await AsyncStorage.getItem(FV_FEATURE_FLAGS);
  return flags || '';
};

// TUTORIALS
export const getCompletedTutorials = async () => {
  const completedTutorials = await AsyncStorage.getItem(FV_COMPLETED_TUTORIALS);
  return completedTutorials ?? '';
};

export const setCompletedTutorials = async (tutorial: string) => {
  const completedTutorials = await getCompletedTutorials();
  if (completedTutorials !== ';') {
    const newListCompletedTutorials = completedTutorials.split(';');
    const uniqueStrings = addNoRepeatedString(newListCompletedTutorials, tutorial);
    await AsyncStorage.setItem(FV_COMPLETED_TUTORIALS, mergArrayIntoStrings(uniqueStrings));
  }
};

export const getLastTimeUserSawTutorial = () =>
  AsyncStorage.getItem(FV_LAST_TIME_USER_SAW_TUTORIAL);

export const setLastTimeUserSawTutorial = () => {
  const isoStringDate = getISOStringDate(new Date());
  AsyncStorage.setItem(FV_LAST_TIME_USER_SAW_TUTORIAL, isoStringDate);
};

export const getFirstTutoMap = async (visualized?: string) => {
  if (visualized === 'visualized') {
    await AsyncStorage.setItem(FV_FIRST_TUTO_MAP, 'visualized');
  }
  const firstMap = await AsyncStorage.getItem(FV_FIRST_TUTO_MAP);
  return firstMap;
};

// PRIVACY POLICY
export const getAcceptedTermUseDate = () => AsyncStorage.getItem(FV_DATE_ACCEPTED_TERM_USE);

export const setDateAcceptedTermUse = (lastDate: string | undefined) => {
  if (lastDate) {
    return AsyncStorage.setItem(FV_DATE_ACCEPTED_TERM_USE, lastDate);
  }
};

export const getUpdatedTermUseDate = () => AsyncStorage.getItem(FV_DATE_UPDATED_TERM_USE);
export const setDateUpdatedTermUse = (updatedDate: string | undefined) => {
  if (updatedDate) {
    return AsyncStorage.setItem(FV_DATE_UPDATED_TERM_USE, updatedDate);
  }
};
