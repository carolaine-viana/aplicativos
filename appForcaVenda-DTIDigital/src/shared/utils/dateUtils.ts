/**
 * Use this to deal/manipulate dates
 */
import * as dateFNS from 'date-fns';
import pt from 'date-fns/locale/pt-BR';
import { zonedTimeToUtc } from 'date-fns-tz';

interface IDateUtils {
  getTimeZonedDate: (date: Date) => Date;
  formatDatePtBr: (date: Date, format?: string) => string;
  formatDateOnlyPtBr: (date: Date, format?: string) => string;
  formatHour: (date: Date, format?: string) => string;
}

const getTimeZonedDate: IDateUtils['getTimeZonedDate'] = (date: Date) =>
  zonedTimeToUtc(date, 'ptBR');

const formatDateOnlyPtBr: IDateUtils['formatDateOnlyPtBr'] = (
  date: Date,
  format = " dd 'de' MMMM "
) =>
  dateFNS.format(date, format, {
    locale: pt,
  });

const formatHour: IDateUtils['formatHour'] = (date: Date, format = "HH:mm'h'") =>
  dateFNS.format(date, format);

const formatDate = (date: Date, format = 'dd-MM-yyyy') => dateFNS.format(date, format);

const getUnixTime = (date: Date) => dateFNS.getUnixTime(date);
const getDateFromUnix = (dateNumber: number) => dateFNS.fromUnixTime(dateNumber);

const getDateWithTime = (date: Date, h: string | number, m: string | number) => {
  const dateWithTime = dateFNS.setHours(date, Number(h));
  dateWithTime.setMinutes(Number(m || 0));
  return dateWithTime;
};

const getNumberDayOfWeek = (date: Date) => dateFNS.getDay(date);

const getLastMonday = (fromDate: Date) => {
  let lastMondayDate = fromDate;
  while (!dateFNS.isMonday(lastMondayDate)) {
    lastMondayDate = dateFNS.subDays(lastMondayDate, 1);
  }
  return lastMondayDate;
};

const addDays = (fromDate = new Date(), amount = 0) => dateFNS.addDays(fromDate, amount);

const addMinutes = (fromDate: Date | number, amount: number) =>
  dateFNS.addMinutes(fromDate, amount);

const getStartOfMonth = (date: Date) => getTimeZonedDate(dateFNS.startOfMonth(date));

const getEndOfMonth = (date: Date) => getTimeZonedDate(dateFNS.endOfMonth(date));

const mountTokenExpirationDate = (tokenExpirationSeconds: number) => {
  const offSetMinutes = 5;
  const now = new Date();
  const time = addMinutes(now, tokenExpirationSeconds / 60 - offSetMinutes);
  return time;
};

const getYearMonth = (
  date = new Date()
): { monthYearNumber: string; monthYearDescription: string } => ({
  monthYearNumber: dateFNS.format(date, 'yyyyMM'),
  monthYearDescription: dateFNS.format(date, "MMMM' 'yyyy", {
    locale: pt,
  }),
});

const getMonthDescription = (month: number = 0, format = 'MMM') => {
  const year = new Date().getFullYear();
  return dateFNS.format(new Date(Number(year), Number(month)), format, {
    locale: pt,
  });
};

const subtractMonths = (date = new Date(), amountMonths: number) =>
  dateFNS.subMonths(date, amountMonths);

const getISOStringDate = (date: Date) => date.toISOString();

const getDiffHoursFromDate = (fromDate: Date, toDate: Date) => {
  const subtractHours = dateFNS.differenceInHours(fromDate, toDate);
  return subtractHours;
};

const getLastMonthsDate = (
  amountMonths: number,
  includeCurrentMonth: boolean = false,
  formatDateString: string = 'MMMM yyyy'
) => {
  let currentDate = new Date();
  const lastMonthsDates: { stringDate: string; pureDate: Date }[] = [];

  for (let index = amountMonths; index > 0; index--) {
    const lastMonth = subtractMonths(currentDate, 1);
    const formatedDate = dateFNS.format(lastMonth, formatDateString, {
      locale: pt,
    });
    lastMonthsDates.push({ stringDate: formatedDate, pureDate: lastMonth });
    currentDate = lastMonth;
  }

  if (includeCurrentMonth) {
    const today = new Date();
    lastMonthsDates.unshift({
      stringDate: dateFNS.format(today, formatDateString, {
        locale: pt,
      }),
      pureDate: today,
    });
  }

  return lastMonthsDates;
};

export {
  formatDate,
  getUnixTime,
  getDateFromUnix,
  getDateWithTime,
  formatHour,
  formatDateOnlyPtBr,
  getNumberDayOfWeek,
  getLastMonday,
  addDays,
  addMinutes,
  getStartOfMonth,
  getEndOfMonth,
  getTimeZonedDate,
  mountTokenExpirationDate,
  getYearMonth,
  getMonthDescription,
  subtractMonths,
  getDiffHoursFromDate,
  getISOStringDate,
  getLastMonthsDate,
};
