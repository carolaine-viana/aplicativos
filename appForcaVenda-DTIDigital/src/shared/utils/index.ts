export * from './generalUtils';
export * from './dateUtils';
export * from './arrayUtils';
export * from './mappers/mapperCustomerVisit';
