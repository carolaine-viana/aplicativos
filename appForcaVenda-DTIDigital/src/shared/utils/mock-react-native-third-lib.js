import { NativeModules as RNNativeModules } from 'react-native';
jest.mock('react-native-reanimated', () => ({ Transition: {
  Together: 'Together',
  Out: 'Out',
  In: 'In',
  Change: 'Change',
}, Transitioning: {
  View: 'Vew'
} }));


jest.mock('@react-navigation/native');
jest.mock('@react-native-firebase/analytics', () => ({
  analytics: jest.fn(),
}));
jest.mock('react-native-fs', () => {
  return class RNFS {};
});
jest.mock('react-native-share', () => {
  return class Share {};
});
jest.mock('base-64', () => ({
  decode: jest.fn(),
}));
jest.mock('react-native-file-viewer', () => {
  return class RNFS {};
});
jest.mock('launchdarkly-react-native-client-sdk', () => {
  return class LDCliente {};
});
jest.mock('react-native-iphone-x-helper', () => ({ getStatusBarHeight: jest.fn() }));
jest.mock('@react-navigation/material-top-tabs', () => ({
  createMaterialTopTabNavigator: jest.fn(),
}));
jest.mock('@react-native-community/netinfo', () => ({
  useNetInfo: () => ({
    isConnected: true,
  }),
}));

RNNativeModules.UIManager = RNNativeModules.UIManager || {};
RNNativeModules.UIManager.RCTView = RNNativeModules.UIManager.RCTView || {};
RNNativeModules.RNGestureHandlerModule = RNNativeModules.RNGestureHandlerModule || {
  State: { BEGAN: 'BEGAN', FAILED: 'FAILED', ACTIVE: 'ACTIVE', END: 'END' },
  attachGestureHandler: jest.fn(),
  createGestureHandler: jest.fn(),
  dropGestureHandler: jest.fn(),
  updateGestureHandler: jest.fn(),
};
RNNativeModules.PlatformConstants = RNNativeModules.PlatformConstants || {
  forceTouchAvailable: false,
};
