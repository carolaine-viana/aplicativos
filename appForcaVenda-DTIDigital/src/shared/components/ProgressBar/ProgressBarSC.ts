import styled from 'styled-components/native';
import { Animated } from 'react-native';
import { getIndicatorColor } from 'shared/utils';

export interface ProgressBarProps {
  step?: number;
  steps?: number;
  height?: number;
  color?: 'red600' | 'yellow400' | 'green400';
  fulfillment?: 'low' | 'high' | 'complete';
}

export const Container = styled.View`
  justify-content: center;
`;

export const ContainerProgress = styled.View<ProgressBarProps>`
  background-color: ${({ theme }) => theme.palette.gray200};
  height: ${(props) => props.height}px;
  border-radius: ${(props) => props.height}px;
  overflow: hidden;
`;

export const Progress = styled(Animated.View)<ProgressBarProps>`
  height: ${(props) => props.height}px;
  width: 100%;
  position: absolute;
  background-color: ${({ step, color, theme }) =>
    theme.palette[color] || getIndicatorColor(step).color};
  left: 0px;
  top: 0px;
  border-radius: ${(props) => props.height}px;
`;
