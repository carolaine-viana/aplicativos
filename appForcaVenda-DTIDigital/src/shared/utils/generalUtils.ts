/**
 * Use this for general utility functions
 */

import { AsyncThunk } from '@reduxjs/toolkit';
import CryptoJS from 'crypto-js';
import theme from 'shared/constants/theme';
import { IPalette } from 'shared/types/IThemeInterface';
import { TAlphabeticalSort, TOption, TPredefinedFilter } from 'shared/types';
import {
  getLastMonday,
  formatDate,
  subtractMonths,
  getYearMonth,
  formatDateOnlyPtBr,
} from './dateUtils';
import { DeviceFeatures } from '../base';
import { IOperationReducer, IJoinQueryParam, ICordinates, IMarker } from '../types';

export const getOperationStatus = <T extends AsyncThunk<any, any, any>>(
  asyncThunkFunction: T
): IOperationReducer<T> => ({
  run: asyncThunkFunction,
  fulfilled: asyncThunkFunction.fulfilled.toString(),
  pending: asyncThunkFunction.pending.toString(),
  rejected: asyncThunkFunction.rejected.toString(),
});

export const handlePlataforms = ({
  handleException,
  handleAndroid,
  handleIOS,
  handlDefault,
}: {
  handleException: Function;
  handleAndroid?: Function | null;
  handleIOS?: Function | null;
  handlDefault?: Function | null;
}) => {
  try {
    if (DeviceFeatures.isAndroid && handleAndroid) {
      handleAndroid();
    } else if (DeviceFeatures.isIOS && handleIOS) {
      handleIOS();
    }

    if (handlDefault) {
      handlDefault();
    }
  } catch (error) {
    handleException(error);
  }
};

export const getKeyValue =
  <T, U extends keyof T>(key: U) =>
  (obj: T) =>
    obj[key];

export const getPartialObject = (obj: any, keepKeys: string[]) =>
  keepKeys.reduce((objResult: any, currentKey: string) => {
    if (obj[currentKey]) {
      return {
        ...objResult,
        [currentKey]: obj[currentKey],
      };
    }
  }, {});

export const mountJoinQuery = (params: IJoinQueryParam): string => {
  const { leftTable, rightTable } = params;
  const query = `select * from ${leftTable.name}
    inner join ${rightTable.name}
    on ${leftTable.name}.${leftTable.key} = ${rightTable.name}.${rightTable.key};`;
  return query
    .replace(/(\r\n|\n|\r)/gm, '')
    .replace(/ +(?= )/g, '')
    .trim();
};

export function mapCordinatesToMarker<T extends Partial<ICordinates>, U extends keyof T>(
  listWithCordinates: T[],
  identifierKey: U
): IMarker<T[U]>[] {
  try {
    return listWithCordinates?.map((item: T) => ({
      latitude: Number(item.latitude),
      longitude: Number(item.longitude),
      identifier: item[identifierKey],
    }));
  } catch {
    return [];
  }
}

export const searchingOptions = (text: string, data: TOption[]) => {
  if (text) {
    return data.filter((item: TOption) => {
      const name = item?.label?.toUpperCase();
      const searchData = text.toUpperCase();
      return name?.includes(searchData);
    });
  }
  return data;
};

export const getCurrency = (value: number = 0) =>
  new Intl.NumberFormat('pt-BR', { style: 'currency', currency: 'BRL' }).format(value);

export const getIndicatorColor = (
  value = 0
): { color: string; textColor: 'error' | 'alert' | 'success' } => {
  if (value <= 90) {
    return {
      color: theme.palette.red600,
      textColor: 'error',
    };
  }
  if (value < 100) {
    return {
      color: theme.palette.yellow400,
      textColor: 'alert',
    };
  }
  return {
    color: theme.palette.green400,
    textColor: 'success',
  };
};

export const getPercentageValue = (value?: number) => {
  if (value) {
    if (value > 0) {
      return value.toFixed(2);
    }
    return value.toFixed(0);
  }
  return 0;
};

export const getNameLetters = (name: string) => {
  const splitedName = name.split(' ');
  const firstName = splitedName[0];
  let lastName = '';
  if (splitedName.length > 1) {
    lastName = splitedName.pop() as string;
    return `${firstName[0].toUpperCase()}${lastName[0].toUpperCase()}`;
  }
  return firstName[0].toUpperCase();
};

export function formatCNPJ(cnpj?: string) {
  if (!cnpj) return '';
  return cnpj.replace(/^(\d{2})(\d{3})(\d{3})(\d{4})(\d{2})/, '$1.$2.$3/$4-$5');
}

export const isTokenExpired = (expirationDate: Date | number) => {
  const now = new Date();
  return now > expirationDate || expirationDate === 0;
};

export const isValidCordinates = (lat: number, lng: number) => lat !== 0 && lng !== 0;

export const mountMonthParam = (date = new Date()) => {
  const month = date.getMonth();
  const monthText = `0${month + 1}`.slice(-2);
  const year = date.getFullYear();
  return `month=${year}${monthText}`;
};

export const mountWeekParam = (date = new Date()) => {
  const lastMonday = getLastMonday(date);
  const formatedDate = formatDate(lastMonday, 'yyyy-MM-dd');
  return `week=${formatedDate}`;
};

const getNumberOfRows = (dataLength: number, numberOfColumns: number) =>
  Math.ceil(dataLength / numberOfColumns);

const sliceData = (data: Array<unknown>, indexStart: number, indexEnd: number) =>
  data.slice(indexStart, indexEnd);

const insertEmptyColumn = (columnsOfRows: Array<unknown>) =>
  columnsOfRows.push({
    text: 'empty',
  });

export function createRowsWithColumns(data, numberOfColumns) {
  const numberOfRowsToCreate = getNumberOfRows(data.length, numberOfColumns);

  let rowsFullFilled = 0;
  let indexStartSlice = 0;
  const listOfRows = [];

  while (rowsFullFilled < numberOfRowsToCreate) {
    const endIndexSlice = indexStartSlice + numberOfColumns;
    const columnsOfRows = sliceData(data, indexStartSlice, endIndexSlice);

    indexStartSlice = endIndexSlice;

    if (columnsOfRows.length < numberOfColumns) {
      insertEmptyColumn(columnsOfRows);
    }

    listOfRows.push(columnsOfRows);
    rowsFullFilled += 1;
  }

  return listOfRows;
}

export function mountChipOpbject<T extends Partial<{ activeColor: keyof IPalette }>>(
  data: Array<T | string>,
  activeColor = 'blue600'
) {
  return data?.map((item: T | string) => {
    if (typeof item === 'string') {
      return { text: item, disabled: false, activeColor };
    }

    return { ...item, activeColor: item.activeColor || activeColor };
  });
}

export const getArrayFromString = (str: string, separator: string = ';') =>
  str.split(separator) || [];

export const mergeArrayStringsWithSeparator = (values: string[], separator: string) =>
  values.join(separator);

export const arrayContainsString = (listOfStrings: string[], stringCheck: string) =>
  listOfStrings.includes(stringCheck);

export const checkArrayIsSubSet = (originalArray: string[], subSetArray: string[]) =>
  subSetArray.every((str) => originalArray.includes(str));

export const mount12MonthsLess = () => {
  const listMonths = [];
  let newDate;
  let newListMonth;

  for (let index = 0; index < 12; index++) {
    newDate = subtractMonths(new Date(), index);
    const { monthYearDescription, monthYearNumber } = getYearMonth(newDate);
    newListMonth = { label: monthYearDescription, value: monthYearNumber };
    listMonths.push(newListMonth);
  }

  return listMonths;
};

export const getMonthYearDescription = (monthYear: string | Date) => {
  if (typeof monthYear === 'string') {
    const [m, y] = monthYear.split(' ');
    return `${m.slice(0, 3).toUpperCase()} / ${y}`;
  }
  const copyDateParam = monthYear as unknown as Date;
  const ptBRDate = formatDateOnlyPtBr(copyDateParam, "MMMM' 'yyyy");
  const [m, y] = ptBRDate.split(' ');
  return `${m.slice(0, 3).toUpperCase()} / ${y}`;
};

export const objectMatchText = <T extends object>(obj: T, searchText = '') => {
  const valuesMerged = Object.values(obj)
    .filter((value) => typeof value === 'number' || typeof value === 'string')
    .join(' ');
  const normalizedString = valuesMerged.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
  const normalizedSearchText = searchText.normalize('NFD').replace(/[\u0300-\u036f]/g, '');
  return normalizedString.toUpperCase().includes(normalizedSearchText.toUpperCase());
};

export const getMonthWeekTabs = () => ({
  week: {
    id: 'week',
    title: 'Semana',
  },
  month: {
    id: 'month',
    title: 'Mês',
  },
});

export const getFilterReceiptTabs = () => ({
  todos: {
    id: 'todos',
    title: 'Todos',
  },
  credito: {
    id: 'credito',
    title: 'Créditos',
  },
  debito: {
    id: 'debito',
    title: 'Débitos',
  },
});
export const filterListByPreDefinedFilters = <T>(
  filters: TPredefinedFilter[],
  listItensToFilter: T[],
  searchStrategy: 'union' | 'intersection' = 'intersection'
) => {
  const arrayMethod = searchStrategy === 'union' ? 'some' : 'every';

  return listItensToFilter?.filter((item) => {
    const keysOfObjetctItem = Object.keys(item);
    const onlyKeysToFilter = filters.filter((filter) => keysOfObjetctItem.includes(filter.prop));

    return onlyKeysToFilter[arrayMethod]((filterCheck) => {
      if (filterCheck.type === 'string') {
        return String(item[filterCheck.prop as keyof T]).includes(filterCheck.value);
      }
      if (filterCheck.type === 'boolean') {
        return !!item[filterCheck.prop as keyof T] === !!filterCheck.shouldPass;
      }
    });
  });
};

export const sortListByKey = <T>(list: T[], key: keyof T, sortMode: TAlphabeticalSort) =>
  [...list]?.sort((a, b) => {
    if (a[key] && b[key]) {
      if (sortMode === 'az') {
        return a[key] < b[key] ? -1 : 1;
      }
      if (sortMode === 'za') {
        return a[key] < b[key] ? 1 : -1;
      }
    }
    return 0;
  });

export const capitalizeText = (text: string) => text && text[0].toUpperCase() + text.slice(1);

export const isValidValue = (value: string | number | boolean | undefined | null) =>
  value !== null && value !== undefined;

export const replaceMultiCharacters = (
  text: string,
  dictionary: Array<{ from: string; to: string }>
) =>
  Array.from(text)
    .map((c) => {
      const foundInDictionary = dictionary.find(
        (dictionaryCharacter) => dictionaryCharacter.from === c
      );
      if (foundInDictionary) {
        return foundInDictionary.to;
      }
      return c;
    })
    .join('');

export const encryptTextTripleDES = (text: string, iv: string, cryptkey: string) => {
  const encrypted = CryptoJS.TripleDES.encrypt(text, CryptoJS.enc.Utf8.parse(cryptkey), {
    iv: CryptoJS.enc.Utf8.parse(iv),
  });
  return encrypted.toString();
};
