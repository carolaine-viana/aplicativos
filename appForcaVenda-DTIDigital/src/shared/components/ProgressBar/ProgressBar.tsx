import React, { useState, useRef, useEffect } from 'react';
import { Animated } from 'react-native';

import { Container, ProgressBarProps, ContainerProgress, Progress } from './ProgressBarSC';

const fulfillmentStatus = {
  low: 'red600',
  high: 'yellow400',
  complete: 'green400',
};

const ProgressBar: React.FC<ProgressBarProps> = ({
  steps = 0,
  step = 0,
  height = 16,
  fulfillment,
  color,
}) => {
  const animatedValue = useRef(new Animated.Value(-1000)).current;
  const reactive = useRef(new Animated.Value(-1000)).current;
  const [width, setWidth] = useState(0);

  useEffect(() => {
    Animated.timing(animatedValue, {
      toValue: reactive,
      duration: 500,
      useNativeDriver: false,
      delay: 300,
    }).start();

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  useEffect(() => {
    const stepMax = step > 100 ? 100 : step;
    reactive.setValue(-width + (width * stepMax) / steps);
  }, [step, width, reactive, steps]);

  return (
    <Container>
      <ContainerProgress height={height}>
        <Progress
          color={fulfillmentStatus[fulfillment?.toLowerCase()] || color}
          height={height}
          step={step}
          onLayout={(e) => {
            const newWidth = e.nativeEvent.layout.width;
            setWidth(newWidth);
          }}
          style={{
            transform: [
              {
                translateX: animatedValue,
              },
            ],
          }}
        />
      </ContainerProgress>
    </Container>
  );
};

export default ProgressBar;
