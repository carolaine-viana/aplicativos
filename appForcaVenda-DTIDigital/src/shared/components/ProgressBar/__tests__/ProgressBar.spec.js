import React from 'react';
import jestExpect from 'expect';
import { render, waitFor, fireEvent } from '@shared/utils/test-utils';
import ProgressBar from '../ProgressBar';

describe('ProgressBar', () => {
  it('progressbar snapshot', () => {
    const screen = render(<ProgressBar />).toJSON();
    jestExpect(screen).toMatchSnapshot();
  });
});
