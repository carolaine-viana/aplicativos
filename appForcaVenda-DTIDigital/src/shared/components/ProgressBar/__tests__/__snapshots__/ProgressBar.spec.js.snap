// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar progressbar snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "justifyContent": "center",
      },
    ]
  }
>
  <View
    height={16}
    style={
      Array [
        Object {
          "backgroundColor": "#D3D5DA",
          "borderRadius": 16,
          "height": 16,
          "overflow": "hidden",
        },
      ]
    }
  >
    <View
      collapsable={false}
      height={16}
      nativeID="animatedComponent"
      onLayout={[Function]}
      step={0}
      style={
        Object {
          "backgroundColor": "#E73933",
          "borderRadius": 16,
          "height": 16,
          "left": 0,
          "position": "absolute",
          "top": 0,
          "transform": Array [
            Object {
              "translateX": -1000,
            },
          ],
          "width": "100%",
        }
      }
    />
  </View>
</View>
`;
